src/Command/PurgeCommand.php

```diff
@@ -0,0 +1,359 @@
+     */
+    private $db;
+
+    /**
+     * @var string
+     */
+    private $rootDir;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var Filesystem
+     */
+    private $fs;
+
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+
+    /**
+     * PurgeCommand constructor.
+     * @param ContaoFrameworkInterface $framework
+     * @param Connection $db
+     * @param string $rootDir
+     * @param LoggerInterface $logger
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param Filesystem|null $fs
+     */
+    public function __construct(
+        ContaoFrameworkInterface $framework,
+        Connection $db,
+        string $rootDir,
+        LoggerInterface $logger,
+        EventDispatcherInterface $eventDispatcher,
+        Filesystem $fs = null
+    ) {
+        $this->framework = $framework;
+        $this->db = $db;
+        $this->rootDir = $rootDir;
+        $this->logger = $logger;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->fs = $fs ? $fs : new Filesystem();
+
+        parent::__construct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('leads:purge')
+            ->setDescription('Purge the leads that are older than the configured storage period.');
+    }
+
+    /**
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int|null|void
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $this->framework->initialize();
+
+        $purged = $this->executeBatchPurge();
+
+        if ($purged) {
+            $logMessage = 'The leads have been purged successfully.';
+        } else {
+            $logMessage = 'No leads to purge.';
+        }
+
+        $logLevel = LogLevel::INFO;
+        $this->logger->log($logLevel, $logMessage, array('contao' => new ContaoContext(__METHOD__, $logLevel)));
+        $output->writeln('<info>'.$logMessage.'</info>');
+    }
+
+    /**
+     * @return bool
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    private function executeBatchPurge()
+    {
+        $purged = false;
+        $forms = $this->db->fetchAll(
+            "SELECT id, title, leadPeriod, leadPurgeUploads FROM tl_form WHERE leadPeriod != ''"
+        );
+
+        foreach ($forms as $masterForm) {
+
+            $leadPeriodTime = $this->convertTimePeriodToTime($masterForm['leadPeriod']);
+            if (!empty($leads = $this->getAllLeads($masterForm['id'], $leadPeriodTime))) {
+
+                $leadsData = $this->getAllLeadsData($leads);
+                $uploads = $this->getUploads($leadsData, $masterForm['leadPurgeUploads']);
+
+                // Add custom logic or modify data before purge
+                $purgeEvent = new LeadsPurgeEvent($masterForm, $leads, $leadsData, $uploads);
```

Comment:
It looks nice to have all leads data available in the event, and have an event listener be able to work with it. However, I wonder if this is really always necessary as it will use quite some memory for the rare case that anyone needs the data. Or do you have any concrete example why this would be necessary?

Also, do we really want the event to be able to _stop_ deleting some records? Because then the users' preferences of deleting data would not be taken care of.

Reply:
you are absolutely right - it was just some playground and in reality it wont make sense to manipulate the data at this point!

perhabs the event should really only dispath after purging with the counts of deleted data - because our customer just wanted some special logging here...